File: packages/python/elizaos/streaming_context.py
"""Streaming context management for automatic streaming in use_model calls.

Follows the OpenTelemetry ContextManager pattern (matching TypeScript implementation):
- Interface for context management
- Uses contextvars for async-safe context propagation
- Automatic streaming in nested async calls

This provides parity with TypeScript's streaming-context.ts.
"""

from collections.abc import Callable
from contextvars import ContextVar
from dataclasses import dataclass
from typing import Any, TypeVar

# Type for generic return values
T = TypeVar("T")


@dataclass
class StreamingContext:
    """Streaming context containing callbacks for streaming lifecycle.

    Attributes:
        on_stream_chunk: Called for each chunk of streamed content.
        on_stream_end: Called when a use_model streaming call completes.
        message_id: Optional message ID for tracking.
        abort_signal: Optional callable returning True if cancelled.
    """

    on_stream_chunk: Callable[[str, str | None], Any]
    """Called for each chunk of streamed content (chunk, message_id) -> None"""

    on_stream_end: Callable[[], None] | None = None
    """Called when a use_model streaming call completes (allows reset between calls)"""

    message_id: str | None = None
    """Optional message ID for tracking"""

    abort_signal: Callable[[], bool] | None = None
    """Optional abort signal - callable returning True if cancelled"""


# Context variable for async-safe streaming context propagation
_streaming_context: ContextVar[StreamingContext | None] = ContextVar(
    "streaming_context", default=None
)


def run_with_streaming_context(
    context: StreamingContext | None,
    fn: Callable[[], T],
) -> T:
    """Run a function with a streaming context.

    All use_model calls within this function will automatically use streaming
    if a context with on_stream_chunk is provided.

    Example:
        async def handle_request(message: Memory) -> None:
            async def on_chunk(chunk: str, msg_id: str | None) -> None:
                await send_sse(chunk)

            ctx = StreamingContext(on_stream_chunk=on_chunk)
            await run_with_streaming_context(
                ctx,
                lambda: runtime.process_message(message)
            )

    Args:
        context: The streaming context with on_stream_chunk callback.
        fn: The function to run with streaming context.

    Returns:
        The result of the function.
    """
    token = _streaming_context.set(context)
    try:
        return fn()
    finally:
        _streaming_context.reset(token)


async def run_with_streaming_context_async(
    context: StreamingContext | None,
    fn: Callable[[], Any],
) -> Any:
    """Run an async function with a streaming context.

    Same as run_with_streaming_context but for async functions.

    Example:
        async def handle_request(message: Memory) -> None:
            async def on_chunk(chunk: str, msg_id: str | None) -> None:
                await send_sse(chunk)

            ctx = StreamingContext(on_stream_chunk=on_chunk)
            await run_with_streaming_context_async(
                ctx,
                lambda: runtime.process_message(message)
            )

    Args:
        context: The streaming context with on_stream_chunk callback.
        fn: The async function to run with streaming context.

    Returns:
        The result of the async function.
    """
    token = _streaming_context.set(context)
    try:
        result = fn()
        # Handle both sync and async functions
        if hasattr(result, "__await__"):
            return await result
        return result
    finally:
        _streaming_context.reset(token)


def get_streaming_context() -> StreamingContext | None:
    """Get the currently active streaming context.

    Called by use_model to check if automatic streaming should be enabled.

    Returns:
        The current streaming context or None.
    """
    return _streaming_context.get()


def set_streaming_context(context: StreamingContext | None) -> None:
    """Set the streaming context directly.

    This is useful for frameworks that manage their own context lifecycle.
    Prefer using run_with_streaming_context when possible.

    Args:
        context: The streaming context to set, or None to clear.
    """
    _streaming_context.set(context)


def clear_streaming_context() -> None:
    """Clear the streaming context.

    This is useful for cleanup after streaming operations.
    """
    _streaming_context.set(None)


__all__ = [
    "StreamingContext",
    "run_with_streaming_context",
    "run_with_streaming_context_async",
    "get_streaming_context",
    "set_streaming_context",
    "clear_streaming_context",
]

File: packages/typescript/src/streaming-context.ts
/**
 * Streaming context management for automatic streaming in useModel calls.
 *
 * Follows the OpenTelemetry ContextManager pattern:
 * - Interface for context management
 * - Platform-specific implementations (Node.js AsyncLocalStorage, Browser Stack)
 * - Auto-detected at runtime - no separate entry points needed
 *
 * @see https://opentelemetry.io/docs/languages/js/context/
 */

import type { StreamChunkCallback } from "./types/components";

/**
 * Streaming context containing callbacks for streaming lifecycle.
 */
export interface StreamingContext {
	/** Called for each chunk of streamed content */
	onStreamChunk: StreamChunkCallback;
	/** Called when a useModel streaming call completes (allows reset between calls) */
	onStreamEnd?: () => void;
	messageId?: string;
	/** Optional abort signal to cancel streaming */
	abortSignal?: AbortSignal;
}

/**
 * Interface for streaming context managers.
 * Different implementations exist for Node.js (AsyncLocalStorage) and Browser (Stack).
 */
export interface IStreamingContextManager {
	/**
	 * Run a function with a streaming context.
	 * The context will be available to all nested async calls via `active()`.
	 */
	run<T>(context: StreamingContext | undefined, fn: () => T): T;

	/**
	 * Get the currently active streaming context.
	 * Returns undefined if no context is active.
	 */
	active(): StreamingContext | undefined;
}

/**
 * Stack-based context manager for browser environments.
 * Safe because browser typically has 1 runtime per request.
 * Supports nested contexts via stack push/pop.
 */
class StackContextManager implements IStreamingContextManager {
	private stack: Array<StreamingContext | undefined> = [];

	run<T>(context: StreamingContext | undefined, fn: () => T): T {
		this.stack.push(context);
		try {
			return fn();
		} finally {
			this.stack.pop();
		}
	}

	active(): StreamingContext | undefined {
		return this.stack.length > 0
			? this.stack[this.stack.length - 1]
			: undefined;
	}
}

// Global singleton - auto-configured on first access
let globalContextManager: IStreamingContextManager | null = null;

function isNodeEnvironment(): boolean {
	return (
		typeof process !== "undefined" &&
		typeof process.versions !== "undefined" &&
		typeof process.versions.node !== "undefined"
	);
}

// Initialize synchronously to avoid the race where early calls use the
// StackContextManager fallback (which doesn't propagate through async/await).
function initContextManagerSync(): IStreamingContextManager {
	if (isNodeEnvironment()) {
		try {
			// eslint-disable-next-line @typescript-eslint/no-require-imports
			const { AsyncLocalStorage } =
				require("node:async_hooks") as typeof import("node:async_hooks");
			const storage = new AsyncLocalStorage<StreamingContext | undefined>();
			return {
				run<T>(context: StreamingContext | undefined, fn: () => T): T {
					return storage.run(context, fn);
				},
				active(): StreamingContext | undefined {
					return storage.getStore();
				},
			} as IStreamingContextManager;
		} catch {
			// AsyncLocalStorage unavailable — fall back to stack
		}
	}
	return new StackContextManager();
}

function getOrCreateContextManager(): IStreamingContextManager {
	if (!globalContextManager) {
		globalContextManager = initContextManagerSync();
	}
	return globalContextManager;
}

/**
 * Set the global streaming context manager.
 * Can be used to override the auto-detected manager.
 *
 * @param manager - The context manager to use globally
 */
export function setStreamingContextManager(
	manager: IStreamingContextManager,
): void {
	globalContextManager = manager;
}

/**
 * Get the global streaming context manager.
 * Auto-detects and creates the appropriate manager on first access.
 */
export function getStreamingContextManager(): IStreamingContextManager {
	return getOrCreateContextManager();
}

/**
 * Run a function with a streaming context.
 * All useModel calls within this function will automatically use streaming.
 *
 * @example
 * ```typescript
 * await runWithStreamingContext(
 *   { onStreamChunk: async (chunk) => sendSSE(chunk), messageId },
 *   async () => {
 *     // All useModel calls here will stream automatically
 *     await runtime.processMessage(message);
 *   }
 * );
 * ```
 *
 * @param context - The streaming context with onStreamChunk callback
 * @param fn - The function to run with streaming context
 * @returns The result of the function
 */
export function runWithStreamingContext<T>(
	context: StreamingContext | undefined,
	fn: () => T,
): T {
	return getOrCreateContextManager().run(context, fn);
}

/**
 * Get the currently active streaming context.
 * Called by useModel to check if automatic streaming should be enabled.
 *
 * @returns The current streaming context or undefined
 */
export function getStreamingContext(): StreamingContext | undefined {
	return getOrCreateContextManager().active();
}

// ---------------------------------------------------------------------------
// useModel → chunk callback delivery (dedupe `model_stream_chunk` hooks)
// ---------------------------------------------------------------------------
// The same provider chunk is often forwarded from useModel's textStream loop *and* from
// DefaultMessageService. Without a turn-scoped marker, pipeline hooks would run twice per
// token (inflated metrics, duplicate side effects). Node uses AsyncLocalStorage depth so
// nested async work stays scoped; non-Node has no ALS store and returns depth 0 (no skip).
// See docs/PIPELINE_HOOKS.md § "Stream hook dedupe (Node)".

let modelStreamChunkDeliveryDepthStorage:
	| import("node:async_hooks").AsyncLocalStorage<number>
	| null = null;
let modelStreamChunkDeliveryStorageInitialized = false;

function getModelStreamChunkDeliveryStorage():
	| import("node:async_hooks").AsyncLocalStorage<number>
	| null {
	if (!modelStreamChunkDeliveryStorageInitialized) {
		modelStreamChunkDeliveryStorageInitialized = true;
		if (isNodeEnvironment()) {
			try {
				// eslint-disable-next-line @typescript-eslint/no-require-imports
				const { AsyncLocalStorage } =
					require("node:async_hooks") as typeof import("node:async_hooks");
				modelStreamChunkDeliveryDepthStorage = new AsyncLocalStorage();
			} catch {
				modelStreamChunkDeliveryDepthStorage = null;
			}
		}
	}
	return modelStreamChunkDeliveryDepthStorage;
}

/**
 * While `> 0`, the runtime is inside `useModel`'s delivery of one `textStream` chunk to
 * `paramsChunk` / `ctxChunk` (after `model_stream_chunk` with `source: "use_model"`).
 * `DefaultMessageService` skips its own `model_stream_chunk` (`source: "message_service"`) in
 * this window so the same raw token is not processed twice. Non-Node environments return `0`.
 */
export function getModelStreamChunkDeliveryDepth(): number {
	const s = getModelStreamChunkDeliveryStorage();
	return s?.getStore() ?? 0;
}

/** Wrap `paramsChunk` / `ctxChunk` invocations from `useModel`'s stream loop. */
export function runInsideModelStreamChunkDelivery<T>(
	fn: () => T | Promise<T>,
): T | Promise<T> {
	const s = getModelStreamChunkDeliveryStorage();
	if (!s) {
		return fn();
	}
	const parent = s.getStore() ?? 0;
	return s.run(parent + 1, fn);
}
